pkg/indexer/reorgHandler.go

```diff
@@ -0,0 +1,156 @@
+package indexer
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/xmtp/xmtpd/pkg/blockchain"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+)
+
+type ChainReorgHandler interface {
+	FindReorgPoint(detectedAt uint64) (uint64, []byte, error)
+}
+
+type ReorgHandler struct {
+	ctx     context.Context
+	client  blockchain.ChainClient
+	queries *queries.Queries
+}
+
+var (
+	ErrNoBlocksFound = errors.New("no blocks found")
+	ErrGetBlock      = errors.New("failed to get block")
+)
+
+// TODO: Make this configurable?
+const BLOCK_RANGE_SIZE uint64 = 1000
+
+func NewChainReorgHandler(
+	ctx context.Context,
+	client blockchain.ChainClient,
+	queries *queries.Queries,
+) *ReorgHandler {
+	return &ReorgHandler{
+		ctx:     ctx,
+		client:  client,
+		queries: queries,
+	}
+}
+
+// TODO: When reorg range has been calculated, alert clients (TBD)
+func (r *ReorgHandler) FindReorgPoint(detectedAt uint64) (uint64, []byte, error) {
+	startBlock, endBlock := blockRange(detectedAt)
+
+	for {
+		storedBlocks, err := r.queries.GetBlocksInRange(
+			r.ctx,
+			queries.GetBlocksInRangeParams{
+				StartBlock: startBlock,
+				EndBlock:   endBlock,
+			},
+		)
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return 0, nil, fmt.Errorf("failed to get stored blocks: %w", err)
+		}
+
+		if len(storedBlocks) == 0 || errors.Is(err, sql.ErrNoRows) {
+			if startBlock == 0 {
+				return 0, nil, ErrNoBlocksFound
+			}
+
+			startBlock, endBlock = blockRange(startBlock)
```

Comment:
I'm not sure, but is it not possible to get into an infinite loop if the returned `startBlock` is 0, but there are blocks in the `storedBlocks` results for that range?

I don't know how it would happen in practice in a blockchain, but we might want to have an explicit exit if `startBlock` is 0 and we don't have the reorg block inside the range.

Reply:
In every path we check for `startBlock == 0` and return, I don't see how we could get into an infinite loop.